samza-core/src/test/java/org/apache/samza/zk/TestZkProcessorLatch.java

```diff
@@ -39,256 +28,193 @@
         //latch.countDown(); only one thread counts down
         try {
-          latch.await(100000, TimeUnit.MILLISECONDS);
+          latch.await(30, TimeUnit.SECONDS);
         } catch (TimeoutException e) {
-          Assert.fail("await timed out. " + e.getLocalizedMessage());
+          Assert.fail(String.format("await timed out from  %s - %s", participant1, e.getLocalizedMessage()));
+        } finally {
+          zkUtils.close();
         }
       });
 
-    Future f2 = pool.submit(
-      () -> {
-        Latch latch = coordinationUtils.getLatch(latchSize, latchId);
-        latch.countDown();
-        try {
-          latch.await(100000, TimeUnit.MILLISECONDS);
-        } catch (TimeoutException e) {
-          Assert.fail("await timed out. " + e.getLocalizedMessage());
-        }
-      });
+    Future f2 = pool.submit(getParticipantRunnable(latchSize, latchId, "participant2"));
 
     try {
-      f1.get(30000, TimeUnit.MILLISECONDS);
-      f2.get(30000, TimeUnit.MILLISECONDS);
+      f1.get(30, TimeUnit.SECONDS);
+      f2.get(30, TimeUnit.SECONDS);
     } catch (Exception e) {
       Assert.fail("failed to get future." + e.getLocalizedMessage());
+    } finally {
+      pool.shutdownNow();
     }
-    pool.shutdownNow();
-  }
-
-  @Test
-  public void testNSizeLatch() {
-    System.out.println("Started N");
-    String latchId = "l1";
-    int latchSize = 3;
-
-    ExecutorService pool = Executors.newFixedThreadPool(3);
-    Future f1 = pool.submit(
-      () -> {
-        Latch latch = coordinationUtils.getLatch(latchSize, latchId);
-        latch.countDown();
-        try {
-          latch.await(100000, TimeUnit.MILLISECONDS);
-        } catch (TimeoutException e) {
-          Assert.fail("await timed out " + e.getLocalizedMessage());
-        }
-      });
-    Future f2 = pool.submit(
-      () -> {
-        Latch latch = coordinationUtils.getLatch(latchSize, latchId);
-        latch.countDown();
-        try {
-          latch.await(100000, TimeUnit.MILLISECONDS);
-        } catch (TimeoutException e) {
-          Assert.fail("await timed out. " + e.getLocalizedMessage());
-        }
-      });
-    Future f3 = pool.submit(
-      () -> {
-        Latch latch = coordinationUtils.getLatch(latchSize, latchId);
-        latch.countDown();
-        try {
-          latch.await(100000, TimeUnit.MILLISECONDS);
-        } catch (TimeoutException e) {
-          Assert.fail("await timed out. " + e.getLocalizedMessage());
-        }
-      });
-
     try {
-      f1.get(300, TimeUnit.MILLISECONDS);
-      f2.get(300, TimeUnit.MILLISECONDS);
-      f3.get(300, TimeUnit.MILLISECONDS);
+      List<String> latchParticipants =
+          testZkUtils.getZkClient().getChildren(
+              String.format("%s/%s_%s", KEY_BUILDER.getRootPath(), ZkProcessorLatch.LATCH_PATH, latchId));
+      Assert.assertNotNull(latchParticipants);
+      Assert.assertEquals(1, latchParticipants.size());
```

Comment:
@nickpan47 :) I didn't even know that a latch size can be different from the number of participating members. 
"testLatchSizeOneWithTwoParticipants" creates a latch of size 1. When there are 2 participants, one of them has to countdown. That is what is being tested. 
This is kind of weird because latch size should be indicative of how many participants we expect to join. Not sure how useful this behavior is.